src/services/ui/src/Router.tsx

```diff
@@ -0,0 +1,58 @@
+import {
+  DefaultGenerics,
+  Outlet,
+  ReactLocation,
+  Route,
+  Router,
+} from "@tanstack/react-location";
+import { Home } from "./pages/Home";
+import { Posts } from "./pages/Posts";
+import { Header } from "./components/Header";
+import { useQueryClient } from "@tanstack/react-query";
+import { getPosts } from "./hooks/queries/useGetPosts";
+import { Post } from "./pages/Post";
+import { getPostById } from "./hooks/queries/useGetPostById";
+
+const location = new ReactLocation();
+
+export const AppRouter = () => {
+  const queryClient = useQueryClient();
+
+  const routes: Route<DefaultGenerics>[] = [
+    {
+      path: "/",
+      element: <Home />,
+    },
+    {
+      path: "/posts",
+      element: <Posts />,
+      loader: () =>
+        queryClient.getQueryData(["posts"]) ??
+        queryClient.fetchQuery(["posts"], getPosts),
+      children: [
+        {
+          path: "/",
+          element: <h1>Select a post</h1>,
+        },
+        {
+          path: ":postId",
+          loader: ({ params: { postId } }) =>
+            queryClient.getQueryData(["posts", postId]) ??
+            queryClient.fetchQuery(["posts", postId], () =>
+              getPostById(postId)
```

Comment:
Correct. So in these loaders you have to make calls that are asynchronous and essentially tell it what to look for. Lower level react query stuff is required here to integrate it like that. In the component itself when it goes to run useQuery it will likely pull from the cache that this loader established

Reply:
oh i see. very cool